mantidimaging/core/io/filenames.py

```diff
@@ -0,0 +1,113 @@
+        else:
+            # Note: allow extra leading digits, for data sets that go 001 ... 998, 999, 1000, 1001
+            self.re_pattern = re.compile("^" + re.escape(prefix) + "([1-9]?[0-9]{" + str(digit_count) + "})" +
+                                         re.escape(suffix) + "$")
+
+        self.re_pattern_metadata = re.compile("^" + re.escape(prefix.rstrip("_ ")) + ".json")
+        self.template = prefix + "{:0" + str(digit_count) + "d}" + suffix
+
+    @classmethod
+    def from_name(cls, filename: str) -> "FilenamePattern":
+        result = FilenamePattern.PATTERN.search(filename)
+
+        if result is None:
+            if "." in filename:
+                name, _, ext = filename.rpartition(".")
+                return FilenamePattern(name, 0, "." + ext)
+            else:
+                return FilenamePattern(filename, 0, "")
+
+        prefix = result.group(1)
+        digits = result.group(2)
+        ext = result.group(3)
+
+        return FilenamePattern(prefix, len(digits), ext)
+
+    def generate(self, index: int) -> str:
+        if self.digit_count == 0:
+            return self.prefix + self.suffix
+        return self.template.format(index)
+
+    def match(self, filename: str) -> bool:
+        return self.re_pattern.match(filename) is not None
+
+    def get_value(self, filename: str) -> int:
+        result = self.re_pattern.match(filename)
+        if result is None:
+            raise ValueError(f"Filename ({filename}) does not match pattern: {self.re_pattern}")
+        if self.digit_count == 0:
+            return 0
+        return int(result.group(1))
+
+    def match_metadata(self, filename: str) -> bool:
+        return self.re_pattern_metadata.match(filename) is not None
+
+
+class FilenameGroup:
+    def __init__(self, directory: Path, pattern: FilenamePattern, all_indexes: List[int]):
+        self.directory = directory
+        self.pattern = pattern
+        self.all_indexes = all_indexes
+        self.metadata_path: Optional[Path] = None
+        self.log_path: Optional[Path] = None
+
+    @classmethod
+    def from_file(cls, path: Path) -> "FilenameGroup":
+        if path.is_dir():
+            raise ValueError(f"path is a directory: {path}")
+        directory = path.parent
+        name = path.name
+        pattern = FilenamePattern.from_name(name)
+        index = pattern.get_value(name)
+        new_filename_pattern = cls(directory, pattern, [index])
+
+        return new_filename_pattern
+
+    def all_files(self) -> Iterator[Path]:
+        for index in self.all_indexes:
+            yield self.directory / self.pattern.generate(index)
+
+    def find_all_files(self) -> None:
+        for filename in self.directory.iterdir():
+            if self.pattern.match(filename.name):
+                self.all_indexes.append(self.pattern.get_value(filename.name))
+
+            if self.pattern.match_metadata(filename.name):
+                self.metadata_path = filename
```

Comment:
Just to check, is the idea that if there were multiple metadata files in a directory then only the last one found would be used for `self.metadata_path`? Or are we assuming there will only ever be one metadata file?